### PR TITLE
UCP/API: Introduce exported memory handle

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Introduce exported memory handle API.

## Why ?

External memory handle will be used to allow using UCP operation on a memory allocated by another process on same or another process, but registered on the same network device.

## How ?

1. Introduce new value `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH` in `ucp_ep_params_flags_field` enumeration.
2. Introduce new value `UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER` in `ucp_mem_map_params_field` enumeration.
3. Introduce new value `UCP_MEM_MAP_EXPORTED` in enumeration which are used then as flags passed to `ucp_mem_map()`.
4. Introduce new field `exported_memh_buffer` in `ucp_mem_map_params_t` structure.
5. Move comments for each field before the value in `enum ucp_ep_params_flags_field`, `enum ucp_mem_map_params_field`, `ucp_mem_map()` flags enumerations to fit 80 symbols per line requirement.

### CODE REVIEW

**Comment by dmitrygx** [comment]
@brminich @shamisp @yosefe @tonycurtis could you review pls?

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -262,31 +262,33 @@ enum ucp_ep_params_field {
...
+    /**
+     * Using a client-server connection establishment mechanism.
+     * @ref ucs_sock_addr_t sockaddr field must be provided and contain the
+     * address of the remote peer.
+     */
+    UCP_EP_PARAMS_FLAGS_CLIENT_SERVER  = UCS_BIT(0),
+
+    /**
+     * Avoid connecting the endpoint to itself when connecting the endpoint
+     * to the same worker it was created on. Affects protocols which send to
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:274]: which -> that

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -262,31 +262,33 @@ enum ucp_ep_params_field {
...
+     * a particular remote endpoint, for example stream.
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:275]: comma after "example"

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -262,31 +262,33 @@ enum ucp_ep_params_field {
...
+     */
+    UCP_EP_PARAMS_FLAGS_NO_LOOPBACK    = UCS_BIT(1),
+
+    /**
+     * Send client id when connecting to remote socket address as part of the
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:280]: id -> identifier, better to put the full word in the comment(ary)

File: src/ucp/api/ucp.h [hunk 4]
```diff
@@ -262,31 +262,33 @@ enum ucp_ep_params_field {
...
+     * connection request payload. On the remote side value can be obtained
+     * from @ref ucp_conn_request_h using @ref ucp_conn_request_query.
+     */
+    UCP_EP_PARAMS_FLAGS_SEND_CLIENT_ID = UCS_BIT(2),
+
+    /**
+     * Request selection of transports with capability to perform local
+     * operations using a peer's memory buffer and memory handle associated
+     * with it and created with @ref UCP_MEM_MAP_EXPORTED capability.
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:289]: comma after "it" reads better (natural breath point)

File: src/ucp/api/ucp.h [hunk 5]
```diff
@@ -532,19 +546,32 @@ enum ucp_datatype_attr_field {
...
+     * Complete the mapping faster, possibly by not populating the pages in the
+     * mapping up-front, and mapping them later when they are accessed by
+     * communication routines.
+     */
+    UCP_MEM_MAP_NONBLOCK = UCS_BIT(0),
+
+    /**
+     * Identify requirement for allocation, if passed address is not a
+     * null-pointer then it will be used as a hint or direct address for
+     * allocation.
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:559]: Think we can simplify here:

If passed address is not a
null-pointer, then it will be used as a hint or direct address for
allocation.

File: src/ucp/api/ucp.h [hunk 6]
```diff
@@ -532,19 +546,32 @@ enum ucp_datatype_attr_field {
...
+    UCP_MEM_MAP_ALLOCATE = UCS_BIT(1),
+
+    /**
+     * Don't interpret address as a hint: place the mapping at exactly that
+     * address. The address must be a multiple of the page size.
+     */
+    UCP_MEM_MAP_FIXED    = UCS_BIT(2),
+
+    /**
+     * Create a exported memory handle which could be shared with peers to
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:570]: an

Comment (tonycurtis) [inline src/ucp/api/ucp.h:570]: which -> that

File: src/ucp/api/ucp.h [hunk 7]
```diff
@@ -532,19 +546,32 @@ enum ucp_datatype_attr_field {
...
+     * perform their UCP operations on a memory allocated by same or another
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:571]: "a memory" ?

not sure what "same or another OS" is referring to

File: src/ucp/api/ucp.h [hunk 8]
```diff
@@ -1600,6 +1627,15 @@ typedef struct ucp_mem_map_params {
       *    internally.
       */
      ucs_memory_type_t      memory_type;
+
+     /**
+      * Exported memory handle buffer to associate a local memory handle with.
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:1632]: with which to associate

File: src/ucp/api/ucp.h [hunk 9]
```diff
@@ -1600,6 +1627,15 @@ typedef struct ucp_mem_map_params {
...
+      * If it is not set (along with its corresponding bit in the field_mask -
+      * @ref UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER), the
+      * @ref ucp_mem_map routine will consider the memory handle buffer set to
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:1635]: to be set

**Review by tonycurtis — CHANGES_REQUESTED** [review]

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:274]: fixed

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:275]: done

File: src/ucp/api/ucp.h (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:280]: done

File: src/ucp/api/ucp.h (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:289]: done

File: src/ucp/api/ucp.h (hunk 5 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:559]: done

File: src/ucp/api/ucp.h (hunk 6 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:570]: done

Comment (dmitrygx) [reply src/ucp/api/ucp.h:570]: done

File: src/ucp/api/ucp.h (hunk 7 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:571]: rephrased

File: src/ucp/api/ucp.h (hunk 8 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:1632]: done

File: src/ucp/api/ucp.h (hunk 9 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:1635]: done

File: src/ucp/api/ucp.h [hunk 10]
```diff
@@ -1600,6 +1629,15 @@ typedef struct ucp_mem_map_params {
       *    internally.
       */
      ucs_memory_type_t      memory_type;
+
+     /**
+      * Exported memory handle buffer with which a local memory handle should
+      * associated. If it is not set (along with its corresponding bit in the
```
Comment (brminich) [inline src/ucp/api/ucp.h:1635]: `should be associated`

Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

**Comment by dmitrygx** [comment]
@brminich @tonycurtis could you review pls again? your comments were addressed.

File: src/ucp/api/ucp.h [hunk 11]
```diff
@@ -262,31 +262,34 @@ enum ucp_ep_params_field {
...
+     * Avoid connecting the endpoint to itself when connecting the endpoint
+     * to the same worker it was created on. Affects protocols that send to
+     * a particular remote endpoint, for example, stream.
+     */
+    UCP_EP_PARAMS_FLAGS_NO_LOOPBACK    = UCS_BIT(1),
+
+    /**
+     * Send client identifier when connecting to remote socket address as
+     * part of the connection request payload. On the remote side value can
+     * be obtained from @ref ucp_conn_request_h using
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:282]: the value

**Review by tonycurtis — CHANGES_REQUESTED** [review]

File: src/ucp/api/ucp.h (hunk 11 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

**Review by brminich — APPROVED** [review]

**Comment by shamisp** [comment]
looks good to me.

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

File: src/ucp/api/ucp.h [hunk 12]
```diff
@@ -262,31 +262,34 @@ enum ucp_ep_params_field {
...
+     * @ref ucp_conn_request_query.
+     */
+    UCP_EP_PARAMS_FLAGS_SEND_CLIENT_ID = UCS_BIT(2),
+
+    /**
+     * Request selection of transports with capability to perform local
+     * operations using a peer's memory buffer and memory handle associated
+     * with it, and created with @ref UCP_MEM_MAP_EXPORTED capability.
+     */
+    UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH  = UCS_BIT(3)
```
Comment (yosefe) [inline src/ucp/api/ucp.h:292]: why need it in ep params?

File: src/ucp/api/ucp.h [hunk 13]
```diff
@@ -532,19 +547,33 @@ enum ucp_datatype_attr_field {
...
+     * Don't interpret address as a hint: place the mapping at exactly that
+     * address. The address must be a multiple of the page size.
+     */
+    UCP_MEM_MAP_FIXED    = UCS_BIT(2),
+
+    /**
+     * Create an exported memory handle that could be shared with peers to
+     * perform their UCP operations on a memory buffer allocated from same
+     * or another virtual memory space, but physically registered on the
+     * same network device.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:574]: maybe add something like `A remote peer should call @ref ucp_mem_map with the flag @ref UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER in order to import and use this memory handle`

File: src/ucp/api/ucp.h (hunk 12 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: to make sure transports that support XGVMI will be selected out of the box.
e.g. if two processes running on the same DPU, they will select SHM, but we need IB transports to perform  operations using imported memory key

File: src/ucp/api/ucp.h (hunk 13 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

File: src/ucp/api/ucp.h (hunk 12 above)
Comment (yosefe) [reply src/ucp/api/ucp.h:292]: what if we always select such transports when running on a DPU?

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: > what if we always select such transports when running on a DPU?

if only they are selected, we couldn't perform xgvmi

Comment (yosefe) [reply src/ucp/api/ucp.h:292]: I mean, what if when running on DPU we always make sure we also select lanes that support xgvmi? so even when both processes run of DPU, they will use it instead of SHM

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: do you mean to somehow understand that we run on DPU/arm and select lanes which support xgvmi w/o requiring to pass `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH` flag?

Comment (yosefe) [reply src/ucp/api/ucp.h:292]: yes

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: but it is not reasonable in the following use cases:
- two processes running on same DPU and perform operations w/o involving HOST memory - better to use SHM;
- two peers running on different DPUs and they have `UCX_CM_USE_ALLDEVICES=n` and no RDMACM can be used - so, UCP won't be able to use TCP.

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: I take the view that it should be decided by user how a specific UCP endpoint is going to be used - for communication w/ DPU system memory (ne need to pass `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH`) or HOST system memory (need to pass `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH`).

Comment (yosefe) [reply src/ucp/api/ucp.h:292]: can we create both types of lanes (for DPU system memory and HOST memory) and use the right one for each operation according to operation parameters?

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: I forgot to mention that `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH` flag is specified for HOST-side endpoints to make sure xgvmi-capable transports will be selected on HOST too.
So, we will need to select lanes according to their characteristics and then add more lanes with xgvmi capability  on HOST and DPU (if they not exist in initial selection).

Comment (yosefe) [reply src/ucp/api/ucp.h:292]: why is it needed on host side? i thought cross-gvmi could work even if you don't have any endpoint between host and DPU and you send the packed exported mkey by any OOB method,

Comment (dmitrygx) [reply src/ucp/api/ucp.h:292]: > why is it needed on host side? i thought cross-gvmi could work even if you don't have any endpoint between host and DPU and you send the packed exported mkey by any OOB method,

right, doesn't make sense to set this flag on HOST side.

if we decided to not have this flag, then we should somehow detect that process created on DPU/arm - and it should be some MD attribute or capability, right?

**Comment by dmitrygx** [comment]
@yosefe @brminich @tonycurtis @shamisp could you review pls?
we moved `UCP_EP_PARAMS_FLAGS_EXPORTED_MEMH` from `ucp_ep_params` to context features (same as `UCP_FEATIRE_AM` and etc). Now the name of the constant is `UCP_FEATURE_EXPORTED_MEMH`.

File: src/ucp/api/ucp.h [hunk 14]
```diff
@@ -138,19 +138,33 @@ enum ucp_params_field {
...
+
+    /** Request Active Message support */
+    UCP_FEATURE_AM            = UCS_BIT(6),  
+
+    /**
+     * Request selection of transports with capability to perform local
+     * operations using a peer's memory buffer and memory handle associated
+     * with it, and created with @ref UCP_MEM_MAP_EXPORTED capability.
+     */
+    UCP_FEATURE_EXPORTED_MEMH = UCS_BIT(3)
```
Comment (brminich) [inline src/ucp/api/ucp.h:167]: `UCS_BIT(7)`

File: src/ucp/api/ucp.h [hunk 15]
```diff
@@ -138,19 +138,33 @@ enum ucp_params_field {
...
+    /** Request stream support */
+    UCP_FEATURE_STREAM        = UCS_BIT(5),
+
+    /** Request Active Message support */
+    UCP_FEATURE_AM            = UCS_BIT(6),  
+
+    /**
+     * Request selection of transports with capability to perform local
+     * operations using a peer's memory buffer and memory handle associated
+     * with it, and created with @ref UCP_MEM_MAP_EXPORTED capability.
```
Comment (brminich) [inline src/ucp/api/ucp.h:165]: i'd rephrase like:
`operations using a peer's memory buffer and memory handle created with @ref UCP_MEM_MAP_EXPORTED capability`

File: src/ucp/api/ucp.h (hunk 14 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:167]: fixed

File: src/ucp/api/ucp.h (hunk 15 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:165]: done

File: src/ucp/api/ucp.h [hunk 16]
```diff
@@ -138,19 +138,33 @@ enum ucp_params_field {
...
-    UCP_FEATURE_AMO32        = UCS_BIT(2),  /**< Request 32-bit atomic
-                                                 operations support */
-    UCP_FEATURE_AMO64        = UCS_BIT(3),  /**< Request 64-bit atomic
-                                                 operations support */
-    UCP_FEATURE_WAKEUP       = UCS_BIT(4),  /**< Request interrupt
-                                                 notification support */
-    UCP_FEATURE_STREAM       = UCS_BIT(5),  /**< Request stream support */
-    UCP_FEATURE_AM           = UCS_BIT(6)   /**< Request Active Message
-                                                 support */
+    /**< Request tag matching support */
```
Comment (yosefe) [inline src/ucp/api/ucp.h:141]: remove `<`

File: src/ucp/api/ucp.h [hunk 17]
```diff
@@ -138,19 +138,33 @@ enum ucp_params_field {
...
+    /** Request stream support */
+    UCP_FEATURE_STREAM        = UCS_BIT(5),
+
+    /** Request Active Message support */
+    UCP_FEATURE_AM            = UCS_BIT(6),  
+
+    /**
+     * Request selection of transports with capability to perform local
+     * operations using a peer's memory buffer and memory handle created
+     * with @ref UCP_MEM_MAP_EXPORTED capability.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:165]: maybe "Support mapping peer's exported memory handle and using it for local operations "

File: src/ucp/api/ucp.h [hunk 18]
```diff
@@ -532,19 +558,35 @@ enum ucp_datatype_attr_field {
...
+
+    /**
+     * Create an exported memory handle that could be shared with peers to
+     * perform their UCP operations on a memory buffer allocated from same
+     * or another virtual memory space, but physically registered on the
+     * same network device. A peer should call @ref ucp_mem_map with the
+     * flag @ref UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER in order to
+     * import and use the memory handle.
+     */
+    UCP_MEM_MAP_EXPORTED = UCS_BIT(3)
```
Comment (yosefe) [inline src/ucp/api/ucp.h:589]: maybe UCP_MEM_MAP_EXPORT ?

File: src/ucp/api/ucp.h [hunk 19]
```diff
@@ -532,19 +558,35 @@ enum ucp_datatype_attr_field {
...
+     */
+    UCP_MEM_MAP_FIXED    = UCS_BIT(2),
+
+    /**
+     * Create an exported memory handle that could be shared with peers to
+     * perform their UCP operations on a memory buffer allocated from same
+     * or another virtual memory space, but physically registered on the
+     * same network device. A peer should call @ref ucp_mem_map with the
+     * flag @ref UCP_MEM_MAP_PARAM_FIELD_EXPORTED_MEMH_BUFFER in order to
+     * import and use the memory handle.
```
Comment (yosefe) [inline src/ucp/api/ucp.h:587]: Create an exported memory handle that could be packed and used by peers for their local operations

File: src/ucp/api/ucp.h [hunk 20]
```diff
@@ -1600,6 +1642,15 @@ typedef struct ucp_mem_map_params {
       *    internally.
       */
      ucs_memory_type_t      memory_type;
+
+     /**
+      * Exported memory handle buffer with which a local memory handle should
```
Comment (yosefe) [inline src/ucp/api/ucp.h:1647]:
```
Exported memory handle buffer as retuned by @ref ... for a memory handle created with @ref .. flag
When this field is specified for @ref ucp_mem_map, the resulting memory handle would be mapping peer memory instead of local memory.
```

File: src/ucp/api/ucp.h (hunk 16 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

File: src/ucp/api/ucp.h (hunk 17 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

File: src/ucp/api/ucp.h (hunk 18 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

File: src/ucp/api/ucp.h (hunk 19 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

File: src/ucp/api/ucp.h (hunk 20 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h]: done

**Comment by dmitrygx** [comment]
@yosefe @brminich @tonycurtis @shamisp could you review pls?

File: src/ucp/api/ucp.h [hunk 21]
```diff
@@ -138,19 +138,33 @@ enum ucp_params_field {
...
+    /** Request stream support */
+    UCP_FEATURE_STREAM        = UCS_BIT(5),
+
+    /** Request Active Message support */
+    UCP_FEATURE_AM            = UCS_BIT(6),  
+
+    /**
+     * Request selection of transports with capability to support mapping a
+     * peer's memory handle created with @ref UCP_MEM_MAP_EXPORT memory mapping
+     * flag and using it for local operations.
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:165]: using -> use

File: src/ucp/api/ucp.h [hunk 22]
```diff
@@ -1600,6 +1642,19 @@ typedef struct ucp_mem_map_params {
...
       */
      ucs_memory_type_t      memory_type;
+
+     /**
+      * Exported memory handle buffer as returned by @ref ucp_mem_map
+      * function for a memory handle created with @ref UCP_MEM_MAP_EXPORT flag.
+      * If this field is specified for @ref ucp_mem_map function, a resulting
+      * memory handle would be a mapping of peer memory instead of local
+      * memory.
+      * If the filed is not set (along with its corresponding bit in the
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:1652]: filed -> field

File: src/ucp/api/ucp.h [hunk 23]
```diff
@@ -532,19 +558,35 @@ enum ucp_datatype_attr_field {
...
+    UCP_MEM_MAP_ALLOCATE = UCS_BIT(1),
+
+    /**
+     * Don't interpret address as a hint: place the mapping at exactly that
+     * address. The address must be a multiple of the page size.
+     */
+    UCP_MEM_MAP_FIXED    = UCS_BIT(2),
+
+    /**
+     * Create an exported memory handle that could be be packed and used by
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:582]: could -> can

File: src/ucp/api/ucp.h [hunk 24]
```diff
@@ -1600,6 +1642,19 @@ typedef struct ucp_mem_map_params {
       *    internally.
       */
      ucs_memory_type_t      memory_type;
+
+     /**
+      * Exported memory handle buffer as returned by @ref ucp_mem_map
+      * function for a memory handle created with @ref UCP_MEM_MAP_EXPORT flag.
+      * If this field is specified for @ref ucp_mem_map function, a resulting
+      * memory handle would be a mapping of peer memory instead of local
```
Comment (tonycurtis) [inline src/ucp/api/ucp.h:1650]: would -> will

**Review by tonycurtis — CHANGES_REQUESTED** [review]

**Review by brminich — APPROVED** [review]

File: src/ucp/api/ucp.h (hunk 21 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:165]: done

File: src/ucp/api/ucp.h (hunk 24 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:1650]: done

File: src/ucp/api/ucp.h (hunk 22 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:1652]: done

**Comment by dmitrygx** [comment]
@tonycurtis @shamisp @brminich @yosefe recent comments have been fixed, could you take a look pls?

**Comment by shamisp** [comment]
Looks good

File: src/ucp/api/ucp.h [hunk 25]
```diff
@@ -160,9 +160,9 @@ enum ucp_feature {
     UCP_FEATURE_AM            = UCS_BIT(6),  
 
     /**
-     * Request selection of transports with capability to perform local
-     * operations using a peer's memory buffer and memory handle created
-     * with @ref UCP_MEM_MAP_EXPORTED capability.
+     * Request selection of transports with capability to support mapping a
```
Comment (yosefe) [inline src/ucp/api/ucp.h:163]: I'd remove "Request selection of transports with capability to.." part

File: src/ucp/api/ucp.h [hunk 26]
```diff
@@ -160,9 +160,9 @@ enum ucp_feature {
...
+     * peer's memory handle created with @ref UCP_MEM_MAP_EXPORT memory mapping
```
Comment (yosefe) [inline src/ucp/api/ucp.h:164]: `Support mapping a peer's memory handle that was created by @ref ucp_mem_map with the flag @ref UCP_MEM_MAP_EXPORT and use it for local operations`

File: src/ucp/api/ucp.h (hunk 25 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:163]: done

File: src/ucp/api/ucp.h (hunk 26 above)
Comment (dmitrygx) [reply src/ucp/api/ucp.h:164]: done

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@tonycurtis @shamisp could you review pls?

**Review by tonycurtis — DISMISSED** [review]

**Review by shamisp — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
